### DISCUSSION TITLE
Displaying GIF animations

### DISCUSSION THREAD

**Discussion by Bodmer** [Bodmer/TFT_eSPI#904] [Compatible support libraries]
The [AnimatedGIF library here](https://github.com/bitbank2/AnimatedGIF) can be used with the TFT_eSPI library.

The "TFT_eSPI_memory" example in the library should work without modification.

**Comment by hai8825** [comment]
Hi @Bodmer So I how or what do I need to modify to get the TFT_eSPI_memory example to display gif files from littleFS or SD? The Adafruit_GFX_SD example from the AnimatedGIF library works for my ILI9341 display but I am trying to achieve the same with your library and ILI9488 display (Pico-ResTouch-LCD-3.5 from waveshare). I got the display to work and display jpg from littleFS but I cant adapt "TFT_eSPI_memory" example to display a gif like Adafruit_GFX_SD example. 

Any help and assistance will be much appreciated it.

Thank you.

**Reply by Bodmer** [reply]
Here is an example for LittleFS:

[LittleFS_GIF.zip](https://github.com/Bodmer/TFT_eSPI/files/12709490/LittleFS_GIF.zip)

You need to set aside memory in FLASH for the LittleFS partition and to upload the gif stored in the sketch data folder.

**Reply by hai8825** [reply]
So I uploaded the gif image that I want to use though the LittleFS data upload. 
![image](https://github.com/Bodmer/TFT_eSPI/assets/78314751/189065a4-74f3-44d7-9d91-deb06144b59f)

However, after uploading the sketch all I get is black screen. I added the following else after the if that open the gif image.
`if (gif.open("/boot_480.gif", GIFOpenFile, GIFCloseFile, GIFReadFile, GIFSeekFile, GIFDraw))
  {
    Serial.printf("Successfully opened GIF; Canvas size = %d x %d\n", gif.getCanvasWidth(), gif.getCanvasHeight());
    tft.startWrite(); // The TFT chip slect is locked low
    while (gif.playFrame(true, NULL))
    {
      yield();
    }
    gif.close();
    tft.endWrite(); // Release TFT chip select for other SPI devices
  }

  else
  {
    Serial.printf("Error opening file = %d\n", gif.getLastError());
    while (1)
    {};
  }`
After re-uploading the sketch, same thing black screen and not the Serial Monitor displays the following messages and it looks like it failed to open the gif image. The question is why was not able to open it and how to fix it,

![image](https://github.com/Bodmer/TFT_eSPI/assets/78314751/83e0f3b6-9cb7-4093-9d8f-14494a249e3d)

**Reply by Bodmer** [reply]
Did the original unmodified sketch work?

**Reply by hai8825** [reply]
The unmodified version of  the Adafruit_GFX_SD example from the AnimatedGIF library works for my ILI9341 display, both reading the image form SD and from LittleFS. I can share that sketch if you would like me tot. The only changes I made were for the display pins and then to use littleFS instead of SD.

**Reply by Bodmer** [reply]
I meant, did the unmodified LittleFS_GIF example I posted work. It worked fine for me. It looks like you changed the file name in your example above.

**Reply by hai8825** [reply]
I tried with your GIF example and tried with mine. It gives the same error for both. "Error opening file.." I will try again just in case but it did not work last time.

EDIT: So I checked again and, yes your image does get displayed and I tried my image resized to 240x 240 instead of 480x320 and it works. The question is why the 480x320 gif cant get opened? The board that I am using is D1 Mini ESP8266.

**Reply by Bodmer** [reply]
It looks like you are using an ESP8266 processor?

I tried my sketch above and it will not compile for the ESP8266! How you got it to compile?

**Reply by hai8825** [reply]
I think I moved the code from GIFDraw sketch to the LittleFS_GIF sketched. I was getting compile error for the File type or something like that. Then I commented put all the code from GIFDraw. I can chare my sketch if it will help.

**Reply by Bodmer** [reply]
I have changed one line so it compiles on an ESP8266:
[LittleFS_GIF.zip](https://github.com/Bodmer/TFT_eSPI/files/12718070/LittleFS_GIF.zip)

This makes the sketch compatible with other processors too.

**Reply by hai8825** [reply]
Yes, I can confirm that the sketch compiles and it displays the PATTERN.GIF image. However, It still fails to display my 480x320 gif. Same result as before. "Error opening file.." Any idea why? Is it the resolution of the gif?

EDIT: I also tried the sketch with D1 Mini ESP32 and it does open your gif and my gif in 240x240 but if it is 480x320 it does not for some reason.

**Reply by hai8825** [reply]
Hi @Bodmer so I think the ESP32, ESP8266 and Raspberry Pico cant handle a gid that is 480x320 on the (Pico-ResTouch-LCD-3.5 from waveshare. Probs due to ram limitations. So for my project I can use a smaller gif like 320x240 which renders just fine with the above example you provided. However I will need it to be centered but not sure how to do that or where the code for it needs to go. Any idea or advice on how to center a GIF image that is smaller than the screen size resolution?
